exonum-java-binding-common/src/test/java/com/exonum/binding/common/proofs/list/ListProofHashCodeCalculatorTest.java

```diff
@@ -0,0 +1,189 @@
+    ListProof root = leafOf(V1);
+
+    calculator = createListProofCalculator();
+    root.accept(calculator);
+
+    assertThat(calculator.getElements(), equalTo(of(0L, V1)));
+    assertEquals(getNodeHashCode(V1), calculator.getCalculatedRootHash());
+  }
+
+  @Test
+  void visit_FullProof2elements() {
+    ListProofElement left = leafOf(V1);
+    ListProofElement right = leafOf(V2);
+    ListProofBranch root = new ListProofBranch(left, right);
+
+    calculator = createListProofCalculator();
+    calculator.visit(root);
+
+    //calculate expected root hash
+    HashCode expectedRootHash = getBranchHashCode(getNodeHashCode(V1), getNodeHashCode(V2));
+
+    assertThat(calculator.getElements(), equalTo(of(0L, V1,
+        1L, V2)));
+    assertEquals(expectedRootHash, calculator.getCalculatedRootHash());
+  }
+
+  @Test
+  void visit_FullProof4elements() {
+    ListProofBranch root = new ListProofBranch(
+        new ListProofBranch(
+            leafOf(V1),
+            leafOf(V2)
+        ),
+        new ListProofBranch(
+            leafOf(V3),
+            leafOf(V4)
+        )
+    );
+
+    //calculate expected root hash
+    HashCode leftBranchHash = getBranchHashCode(getNodeHashCode(V1), getNodeHashCode(V2));
+    HashCode rightBranchHash = getBranchHashCode(getNodeHashCode(V3), getNodeHashCode(V4));
+    HashCode expectedRootHash = getBranchHashCode(leftBranchHash, rightBranchHash);
+
+    calculator = createListProofCalculator();
+    calculator.visit(root);
+
+    assertThat(calculator.getElements(),
+        equalTo(of(0L, V1,
+            1L, V2,
+            2L, V3,
+            3L, V4)));
+    assertEquals(expectedRootHash, calculator.getCalculatedRootHash());
+  }
+
+  @Test
+  void visit_ProofLeftValue() {
+    ListProof left = leafOf(V1);
+    ListProof right = new ListProofHashNode(H2);
+    ListProofBranch root = new ListProofBranch(left, right);
+
+    HashCode expectedRootHash = getBranchHashCode(getNodeHashCode(V1), H2);
+
+    calculator = createListProofCalculator();
+    calculator.visit(root);
+
+    assertThat(calculator.getElements(), equalTo(of(0L, V1)));
+    assertEquals(expectedRootHash, calculator.getCalculatedRootHash());
+  }
+
+  @Test
+  void visit_ProofRightValue() {
+    ListProof left = new ListProofHashNode(H1);
+    ListProof right = leafOf(V2);
+    ListProofBranch root = new ListProofBranch(left, right);
+
+    calculator = createListProofCalculator();
+    calculator.visit(root);
+
+    HashCode expectedRootHash = getBranchHashCode(H1, getNodeHashCode(V2));
+
+    assertThat(calculator.getElements(), equalTo(of(1L, V2)));
+    assertEquals(expectedRootHash, calculator.getCalculatedRootHash());
+  }
+
+  @Test
+  void visit_FullProof3elements() {
+    ListProofBranch root = new ListProofBranch(
+        new ListProofBranch(
+            leafOf(V1),
+            leafOf(V2)
+        ),
+        new ListProofBranch(
+            leafOf(V3),
+            new ListProofHashNode(H1)
```

Comment:
It makes it a 4 element tree, not three. There must be `null`,
indicating that there is no more elements.